src/config/mongoose_config_terms.erl

```diff
@@ -0,0 +1,145 @@
+config_filenames_to_include([_Other | Terms]) ->
+    config_filenames_to_include(Terms);
+config_filenames_to_include([]) ->
+    [].
+
+include_config_files(Terms, Configs) ->
+    include_config_files(Terms, Configs, []).
+
+include_config_files([], _Configs, Res) ->
+    Res;
+include_config_files([{include_config_file, Filename} | Terms], Configs, Res) ->
+    include_config_files([{include_config_file, Filename, []} | Terms],
+                         Configs, Res);
+include_config_files([{include_config_file, Filename, Options} | Terms],
+                     Configs, Res) ->
+    IncludedTerms = find_plain_terms_for_file(Filename, Configs),
+    Disallow = proplists:get_value(disallow, Options, []),
+    IncludedTerms2 = delete_disallowed(Disallow, IncludedTerms),
+    AllowOnly = proplists:get_value(allow_only, Options, all),
+    IncludedTerms3 = keep_only_allowed(AllowOnly, IncludedTerms2),
+    include_config_files(Terms, Configs, Res ++ IncludedTerms3);
+include_config_files([Term | Terms], Configs, Res) ->
+    include_config_files(Terms, Configs, Res ++ [Term]).
+
+find_plain_terms_for_file(Filename, Configs) ->
+    case lists:keyfind(Filename, 1, Configs) of
+        false ->
+            %% Terms were not provided by caller for this file
+            erlang:error({config_not_found, Filename});
+        {Filename, Terms} ->
+            Terms
+    end.
+
+%% @doc Filter from the list of terms the disallowed.
+%% Returns a sublist of Terms without the ones which first element is
+%% included in Disallowed.
+-spec delete_disallowed(Disallowed :: [atom()],
+                        Terms :: [term()]) -> [term()].
+delete_disallowed(Disallowed, Terms) ->
+    lists:foldl(
+      fun(Dis, Ldis) ->
+          delete_disallowed2(Dis, Ldis)
+      end,
+      Terms,
+      Disallowed).
+
+delete_disallowed2(Disallowed, [H | T]) ->
+    case element(1, H) of
+        Disallowed ->
+            ?WARNING_MSG("event=ignore_disallowed_option option=~p", [Disallowed]),
+            delete_disallowed2(Disallowed, T);
+        _ ->
+            [H | delete_disallowed2(Disallowed, T)]
+    end;
+delete_disallowed2(_, []) ->
+    [].
+
+%% @doc Keep from the list only the allowed terms.
+%% Returns a sublist of Terms with only the ones which first element is
+%% included in Allowed.
+-spec keep_only_allowed(Allowed :: [atom()],
+                        Terms :: [term()]) -> [term()].
+keep_only_allowed(all, Terms) ->
+    Terms;
+keep_only_allowed(Allowed, Terms) ->
+    {As, NAs} = lists:partition(
+                  fun(Term) ->
+                      lists:member(element(1, Term), Allowed)
+                  end,
+                  Terms),
+    [?WARNING_MSG("event=ignore_disallowed_option option=~p", [NA])
+     || NA <- NAs],
+    As.
+
+%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
+%%% Errors reading the config file
+
+-type config_problem() :: atom() | {integer(), atom() | tuple(), _}. % spec me better
```

Comment:
I know this is how it was in the `ejabberd_config.erl` but what do you think of moving this type to the top together with the other types.